### PR TITLE
add publish test result

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by Sourcery

This pull request updates the CI workflow by adding permissions for checks and pull-requests, upgrading the Codecov action to v4, and integrating a new step to publish test results using the mikepenz/action-junit-report@v4 action.

- **CI**:
    - Added permissions for checks and pull-requests in the Codecov job.
    - Updated Codecov action from v3 to v4.
    - Integrated 'Publish Test Results' step using mikepenz/action-junit-report@v4 to always publish test results with detailed summary.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `improve_review_guide`.
Run ID: `improve_review_guide/benchmark_2024-08-06T17-38-05_v1-21-0-70-g56484f776-dirty`.

This pull request was cloned from `https://github.com/sourcery-ai-experiments/iptux/pull/5`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      docstrings_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm

# The pull request dataset to run the experiment on
pull_request_dataset:
  # open-source PRs
- https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
- https://github.com/sourcery-ai-experiments/iptux/pull/5
- https://github.com/sourcery-ai-experiments/iptux/pull/6
- https://github.com/sourcery-ai-experiments/kalima/pull/3
- https://github.com/sourcery-ai-experiments/PED/pull/3
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
- https://github.com/sourcery-ai-experiments/Trenova/pull/11
  # Sourcery PRs
- https://github.com/sourcery-ai/core/pull/4055
- https://github.com/sourcery-ai/core/pull/4263
- https://github.com/sourcery-ai/core/pull/4352
- https://github.com/sourcery-ai/core/pull/4364
- https://github.com/sourcery-ai/core/pull/4366
- https://github.com/sourcery-ai/core/pull/4390
- https://github.com/sourcery-ai/core/pull/4402
- https://github.com/sourcery-ai/core/pull/4445
- https://github.com/sourcery-ai/core/pull/4462
- https://github.com/sourcery-ai/core/pull/4463

review_comment_labels: []

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews: []

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request updates the CI workflow by adding permissions for checks and pull-requests, upgrading the Codecov action to v4, and integrating a new step to publish test results using the mikepenz/action-junit-report@v4 action.

### File-Level Changes

| Files | Changes |
| ----- | ------- |
| `.github/workflows/ci.yml` | Updated CI workflow to include permissions for checks and pull-requests, upgraded Codecov action to v4, and added a step to publish test results using mikepenz/action-junit-report@v4. |

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -73,6 +73,9 @@ jobs:
 
   codecov:
     runs-on: ubuntu-22.04
+    permissions:
+      checks: write
+      pull-requests: write
```
Comment (SourceryAI) [inline .github/workflows/ci.yml:78]: **🚨 issue (security):** Review the necessity of write permissions for checks and pull requests.

The added permissions grant write access to checks and pull requests. Please review the documentation for the Codecov and 'Publish Test Results' actions to confirm if these write permissions are necessary. If not, consider limiting the scope to read-only to enhance security.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider adding comments in the `ci.yml` file to explain the rationale behind the new permissions and the addition of the 'Publish Test Results' step. This will help future maintainers understand the changes.
- Ensure that the new permissions for checks and pull-requests are necessary and do not introduce any security risks.
- Run benchmarks to ensure that the added steps do not significantly increase the CI runtime.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟢 **General issues**: all looks good
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/818177f2-32d7-4134-97e4-36c5b8861aea?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**